Change gRPC comparison IPC link, and use consistent title casing (#11744)
aspnet/Docs: aspnetcore/grpc/comparison.md

> ms.date: 03/26/2019

Reviewer: Update the date

uid: grpc/comparison
---
# Comparing gRPC services with HTTP APIs

By [James Newton-King](https://twitter.com/jamesnk)

> This article provides a comparison between [gRPC services](https://grpc.io/docs/guides/) and HTTP APIs, and recommends scenarios for using gRPC over other technologies.

Reviewer: To avoid that comma splice (before the last "and") and avoid the awkward and...and structure, I recommend using the list format for this (we do it quite frequently to break up long sentences for machine translation anyway) ...

> This article provides:
>
> \* A comparison between \[gRPC services](https://grpc.io/docs/guides/) and HTTP APIs.
> \* Recommended scenarios for using gRPC over other technologies.
```

Author: I'd like a more conversational paragraph to introduce the article. I've made some changes.

Reviewer: I generally agree with your sentiment.

* Binary framing and compression. HTTP/2 protocol is compact and efficient both in sending and receiving.
* Multiplexing of multiple HTTP/2 calls over a single TCP connection. Multiplexing eliminates [head-of-line blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking).

### Code generation

All gRPC frameworks provide first-class support for code generation. A core file to gRPC development is the [`*.proto` file](https://developers.google.com/protocol-buffers/docs/proto3), which defines the contact of gRPC services and messages. From this file gRPC frameworks will code generate a service base class, messages, and a complete client.

By sharing the `*.proto` file between the server and client, messages and client code can be generated from end to end. Code generation of the client eliminates duplication of messages on the client and server, and creates a strongly-typed client for you. Not having to write a client saves significant development time in applications with many services.

### Strict specification

> There is no a formal agreement of what an HTTP API with JSON should look like. The lack of an agreement creates debate over the format of URLs, HTTP verbs, and response codes.

Reviewer: ```suggestion
There isn't a formal agreement on a specification for HTTP API with JSON. The lack of an agreement creates debate over the format of URLs, HTTP verbs, and response codes.
```

> The [gRPC specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) is prescriptive about the format a gRPC service must follow. gRPC eliminates debate and saves developer time through its simplicity.

Reviewer: Be careful with "simple" and variants (e.g., "simplicity").

These words are generally **_BANNED_** on :skull: **_pain of death_** :skull: on the repo. :smile: lol ... seriously tho ... for good reason in technical writing. Such remarks can be fleshed out. For example ...

```suggestion
The [gRPC specification](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) is prescriptive about the format a gRPC service must follow. gRPC eliminates debate and saves developer time because gPRC implementations are consistent across platforms and implementations.
```

... a bit more specific.

Propagating the deadline and cancellation through child gRPC calls helps enforce resource usage limits.

## gRPC recommended scenarios

gRPC is well suited to the following scenarios:

* **Microservices** &ndash; gRPC is designed low latency and high throughput communication. gRPC is great for lightweight microservices where efficiency is critical.
* **Point-to-point real-time communication** &ndash; gRPC has excellent support for bi-directional streaming. gRPC services can push messages in real-time without polling.
* **Polygot environments** &ndash; gRPC tooling supports all popular development languages, making gRPC a good choice for multi-language environments.

> * **Network constrained environments** &ndash; gRPC messages are serialized with Protobuf, a lightweight message format. A gRPC message will always be smaller than an equivalent JSON message.

Reviewer: Avoid future tense (and state verbs) ...

```suggestion
* **Network constrained environments** &ndash; gRPC messages are serialized with Protobuf, a lightweight message format. A gRPC message is always smaller than an equivalent JSON message.
```

### Not human readable

HTTP API requests are sent as text and can be read and created by humans.

gRPC messages are encoded with Protobuf by default. While Protobuf is efficient to send and receive, its binary format isn't human readable. Protobuf requires the message's interface description specified in the `*.proto` file to properly deserialize. Additional tooling is required to analyze Protobuf payloads on the wire and to compose requests by hand.

Features such as [server reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) and the [gRPC command line tool](https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) exist to assist with binary Protobuf messages. Also, Protobuf messages support [conversion to and from JSON](https://developers.google.com/protocol-buffers/docs/proto3#json). The built-in JSON conversion provides an efficient way to convert Protobuf messages to and from human readable form when debugging.

## Alternative framework scenarios

Other frameworks are recommended over gRPC in the following scenarios:

> * **Browser accessible APIs** &ndash; gRPC is not fully supported in the browser. gRPC-Web can offer browser support, but it has limitations and introduces a server proxy.

Reviewer: MS style manual always calls for contractions ...

```suggestion
* **Browser accessible APIs** &ndash; gRPC isn't fully supported in the browser. gRPC-Web can offer browser support, but it has limitations and introduces a server proxy.
```

> * **Broadcast real-time communication** &ndash; gRPC supports real-time communication via streaming, but it does not have the concept broadcasting a message out to registered connections. For example, in a chat room scenario where new chat messages should be sent to all clients in the chat room, each gRPC call would need to individually stream new chat messages to the client. [SignalR](xref:signalr/introduction) is a good framework for this scenario. It has the concept of persistent connections and built-in support for broadcasting messages.

Reviewer: Same deal on contractions and state verbs ...

```suggestion
* **Broadcast real-time communication** &ndash; gRPC supports real-time communication via streaming, but the concept of broadcasting a message out to registered connections doesn't exist. For example in a chat room scenario where new chat messages should be sent to all clients in the chat room, each gRPC call is required to individually stream new chat messages to the client. [SignalR](xref:signalr/introduction) is a useful framework for this scenario. SignalR has the concept of persistent connections and built-in support for broadcasting messages.
```

## gRPC weaknesses

### Limited browser support

It's impossible to directly call a gRPC service from a browser today. gRPC heavily uses HTTP/2 features and no browser provides the level of control required over web requests to support a gRPC client. For example, browsers do not allow a caller to require that HTTP/2 be used, or provide access to underlying HTTP/2 frames.

[gRPC-Web](https://grpc.io/docs/tutorials/basic/web.html) is an additional technology from the gRPC team that provides limited gRPC support in the browser. gRPC-Web consists of two parts: a JavaScript client that supports all modern browsers, and a gRPC-Web proxy on the server. The gRPC-Web client calls the proxy and the proxy will forward on the gRPC requests to the gRPC server.

Not all of gRPC's features are supported by gRPC-Web. Client and bi-directional streaming isn't supported, and there is limited support for server streaming.

### Not human readable

HTTP API requests are sent as text and can be read and created by humans.

> gRPC messages are encoded with Protobuf by default. While Protobuf is efficient to send and receive, its binary format is not human readable. Protobuf requires the message's interface description specified in the `*.proto` file to properly deserialize. Additional tooling needs to be used to analyze Protobuf payloads on the wire and to compose requests by hand.

Reviewer: ```suggestion
gRPC messages are encoded with Protobuf by default. While Protobuf is efficient to send and receive, its binary format isn't human readable. Protobuf requires the message's interface description specified in the *.proto* file to properly deserialize. Additional tooling is required to analyze Protobuf payloads on the wire and to compose requests by hand.
```

> Features like [server reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) and [gRPC command line tool](https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) exist to get around this limitation. Also, Protobuf messages support [conversion to and from JSON](https://developers.google.com/protocol-buffers/docs/proto3#json). The built-in JSON conversion provides a good way to convert the protobuf messages to/from human readable form when debugging.

Reviewer: ```suggestion
Features such as [server reflection](https://github.com/grpc/grpc/blob/master/doc/server-reflection.md) and [gRPC command line tool](https://github.com/grpc/grpc/blob/master/doc/command_line_tool.md) exist to assist with binary Protobuf messages. Also, Protobuf messages support [conversion to and from JSON](https://developers.google.com/protocol-buffers/docs/proto3#json). The built-in JSON conversion provides an efficient way to convert Protobuf messages to and from human readable form when debugging.
```

Is this a new paragraph? There's no blank line between lines 92 and 93. It can stand as a new para with the explicit naming of "this limitation" that I provided in the suggestion.

## Alternative framework scenarios

Other frameworks are recommended over gRPC in the following scenarios:

* **Browser accessible APIs** - gRPC is not fully supported in the browser. gRPC-Web can offer browser support, but it has limitations and introduces a server proxy.
* **Broadcast real-time communication** - gRPC supports real-time communication via streaming, but it does not have the concept broadcasting a message out to registered connections. For example, in a chat room scenario where new chat messages should be sent to all clients in the chat room, each gRPC call would need to individually stream new chat messages to the client. [SignalR](xref:signalr/introduction) is a good framework for this scenario. It has the concept of persistent connections and built-in support for broadcasting messages.

> * **Inter-process communication** - A process would need to host a HTTP/2 server to accept incoming gRPC calls. For Windows inter-process communication [pipes](/dotnet/standard/io/pipe-operations) is a fast, lightweight method of communication.

Reviewer: We're using the en dash for these types of lists. Instead of `-`, use `&ndash;` on Lines 99-101.

```suggestion
* **Inter-process communication** &ndash; A process must host an HTTP/2 server to accept incoming gRPC calls. For Windows, inter-process communication [pipes](/dotnet/standard/io/pipe-operations) is a fast, lightweight method of communication.
```
